packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -166,7 +166,8 @@ export default (superClass: Class<Parser>): Class<Parser> =>
         !this.match(tt.colon) &&
         !this.match(tt.eq) &&
         !this.match(tt.question) &&
-        !this.match(tt.bang)
+        !this.match(tt.bang) &&
+        !this.match(tt.relational)
```

Comment:
I think we could align to TS implementation here:

https://github.com/microsoft/TypeScript/blob/4885dec80d184ef1570ac97739679ca66af7181d/src/compiler/parser.ts#L1796

An allowlist instead of denylist is better to reason about and avoid any surprise cases that we did not rule out.